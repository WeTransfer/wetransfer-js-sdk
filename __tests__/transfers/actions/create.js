const routes = require('../../../src/config/routes');
const createAction = require('../../../src/transfers/actions/create');
const WTError = require('../../../src/error');

describe('Create transfer action', () => {
  let create = null;
  const mocks = {};
  beforeEach(() => {
    mocks.request = { send: jest.fn() };
    mocks.uploadFileToTransfer = jest.fn();
    mocks.finalizeTransfer = jest.fn((transfer) => transfer);

    mocks.request.send.mockReturnValue({
      id: 'random-hash',
      message: 'Little kittens',
      state: 'uploading',
      files: [
        {
          id: 'random-hash',
          name: 'kittie.gif',
          size: 1024,
          type: 'file',
        },
      ],
      expires_at: '2018-01-01T00:00:00Z',
    });

    create = createAction({
      routes,
      request: mocks.request,
      uploadFileToTransfer: mocks.uploadFileToTransfer,
      finalizeTransfer: mocks.finalizeTransfer,
    });
  });

  it('should create a create transfer request', async () => {
    await create({
      message: 'WeTransfer SDK',
      files: [
        {
          name: 'kittie.gif',
          size: 1024,
        },
      ],
    });
    expect(mocks.request.send).toHaveBeenCalledWith(
      {
        url: '/v2/transfers',
      },
      {
        message: 'WeTransfer SDK',
        files: [
          {
            name: 'kittie.gif',
            size: 1024,
          },
        ],
      }
    );
  });

  it('should return a RemoteTransfer object', async () => {
    const board = await create({
      message: 'WeTransfer SDK',
      files: [
        {
          name: 'kittie.gif',
          size: 1024,
        },
      ],
    });
    expect(board).toMatchSnapshot();
  });

  it('should create an upload file request if content is provided', async () => {
    const transfer = await create({
      message: 'WeTransfer SDK',
      files: [
        {
          name: 'kittie.gif',
          size: 1024,
          content: [],
        },
      ],
    });
    expect(transfer).toMatchSnapshot();
    expect(mocks.uploadFileToTransfer).toHaveBeenCalledWith(
      transfer,
      {
        id: 'random-hash',
        name: 'kittie.gif',
        size: 1024,
        type: 'file',
        multipart: undefined,
        chunks: [],
      },
      []
    );
  });

  it('should create a finalize transfer request if content is provided', async () => {
    const transfer = await create({
      message: 'WeTransfer SDK',
      files: [
        {
          name: 'kittie.gif',
          size: 1024,
          content: [],
        },
      ],
    });
    expect(transfer).toMatchSnapshot(transfer);
    expect(mocks.finalizeTransfer).toHaveBeenCalledWith(transfer);
  });

  it('should throw an exception if the request fails', async () => {
    mocks.request.send.mockImplementation(() =>
      Promise.reject(new Error('Network error.'))
    );

    try {
      await create({
        name: 'WeTransfer SDK',
      });
      await authorize();
    } catch (error) {
      expect(error).toBeInstanceOf(WTError);
    }
  });
});
