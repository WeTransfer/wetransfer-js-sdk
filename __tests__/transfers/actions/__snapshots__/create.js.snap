// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create transfer action should create a finalize transfer request if content is provided 1`] = `
Object {
  "expires_at": Object {},
  "files": Array [
    Object {
      "id": "random-hash",
      "name": "kittie.gif",
      "size": 1024,
      "type": "file",
    },
  ],
  "id": "random-hash",
  "message": "Little kittens",
  "state": "uploading",
}
`;

exports[`Create transfer action should create an upload file request if content is provided 1`] = `
RemoteTransfer {
  "expires_at": 2018-01-01T00:00:00.000Z,
  "files": Array [
    Object {
      "id": "random-hash",
      "name": "kittie.gif",
      "size": 1024,
      "type": "file",
    },
  ],
  "id": "random-hash",
  "message": "Little kittens",
  "state": "uploading",
}
`;

exports[`Create transfer action should return a RemoteTransfer object 1`] = `
RemoteTransfer {
  "expires_at": 2018-01-01T00:00:00.000Z,
  "files": Array [
    Object {
      "id": "random-hash",
      "name": "kittie.gif",
      "size": 1024,
      "type": "file",
    },
  ],
  "id": "random-hash",
  "message": "Little kittens",
  "state": "uploading",
}
`;
